solidity-vyper-updates/Vyper Online Compiler - Your first smart contract.md

```diff
@@ -0,0 +1,172 @@
+
+If everything is ok, you should see a tick on Bytecode, ABI and LLL tabs, this indicates the compilation was successful.
+
+![](https://api.kauri.io:443/ipfs/QmTLoPwXJ1TooRbC9hagxHcBXkPNAJsFULvCnhfUBzWpdg)
+
+If there is an error, you should be able to see the error message when you open any of the tab links.
+
+![](https://api.kauri.io:443/ipfs/QmcTdx9ifH6nipE2KeaWjkz6EtDMgk3m3ib4zGCX1vyipM)
+
+## Issuing a Bounty
+
+Now that we have the basic skeleton of our smart contract, we can start adding functions, first we will tackle allowing a user to issue a bounty.
+
+### Declare state variables
+
+Just like in solidity, there are state varaibles in vyper. State variables are values which are permanently stored in a contract storage.
+
+[You can see a full list of vyper types in the vyper types documentation ](https://vyper.readthedocs.io/en/latest/types.html)
+
+Next we define events for the contract. Vyper can log events caught during runtime and display it for the user.
+
+```
+BountyIssued: event({_id: int128, _issuer: indexed(address), _amount: wei_value, data: bytes32 })
+```
+
+```
+BountyCancelled: event({ _id: int128, _issuer: indexed(address), _amount: wei_value })
+```
+
+```
+BountyFulfilled: event({ _bountyId: int128, _issuer: indexed(address), _fulfiller: indexed(address), _fulfillmentId: int128, _amount: wei_value})
+```
+
+```
+FulfillmentAccepted: event({ _bountyId: int128, _issuer: indexed(address), _fulfiller: indexed(address), _fulfillmentId: int128, _amount: wei_value })
+```
+
+We have four events that has different fields. The user may wish to listen to the events for changes on the contract.
+
+Let's declare an constant values which we’ll use to keep track of a bounties state
+
+```
+CREATED: constant(uint256) = 0
+ACCEPTED: constant(uint256) = 1
+CANCELLED: constant(uint256) = 2
+```
+
+Now, let's define 2 arrays where we will store data about each issued bounty and fulfillments
+
+```
+bounties: map(int128, Bounty)
+fulfillments: map(int128, Fulfillment)
+```
+
+Before we move to functions, in vyper we have decorators written on top of each of the function. [See functions](https://vyper.readthedocs.io/en/latest/structure-of-a-contract.html#functions)
+
+### Issue Bounty Function
+
+Now that we have declared our state variables we can now add functions to allow users to interact with our smart contract
+
+```
+@public
+@payable
+```
+
+We add the public decorator to the function so that it can be called from the contract. In order to send ETH to our contract we need to add the payable keyword to our function. Without this payable keyword the contract will reject all attempts to send ETH to it via this function. Read more about [decorators](https://vyper.readthedocs.io/en/latest/structure-of-a-contract.html#functions).
+
+```
+def issueBounty(_data: bytes32, _deadline: timestamp):
+    assert msg.value > 0
+    assert _deadline > block.timestamp
+
+    bIndex: int128 = self.nextBountyIndex
```

Comment:
@iamonuwa There were a few code errors, most I managed to fix, but I am stuck on this one. At this line I get:

```
line 34: Persistent variable undeclared: nextBountyIndex bIndex: int128 = self.nextBountyIndex
```

Reply:
I'll take a look

Reply:
@iamonuwa Thanks!